plugin dependency loading

danielepantaleone opened the pull request:
I made some major changes to plugin loading. The `Plugin` class now declares a new attribute customizable by plugin developers: `requiresPlugins`: it's a list of plugins names a plugin needs in order to work properly.
It's fairly simple to use. So if I have 3 plugins `custom1`, `custom2`, `custom3` and `custom1` needs the other 2 to work, the syntax will be:

```python
class Custom1Plugin(b3.plugin.Plugin):
    requiresPlugins = ['custom2', 'custom3']
```
Everything will be handled then by the B3 core.
The parser now follows those steps:

* First load the plugin list from B3 main configuration file
* Then remove from the list of plugins to be loaded the ones that have missing modules (or configuration file if they require one: the automatic configuration file search still works here)
* Check that all the plugins in the plugin list meets the necessary requirements: here B3 checks for plugin dependency. When a dependency is not met, B3 tries to fix this by trying to load the missing plugin (auto search in plugins directories, both module and configuration file). This process happens recursively using DFS (so load plugin dependencies from the ground up). If also one of those dependencies can't be satisfied, the *root* plugin will be discarded and not loaded.
* Topological sort over the plugin list: make sure that plugin object instances are created according to the plugin depencency order (so in the above case `custom2` and `custom3` are initialized before `custom1`)
* Admin plugin still is sorted as first plugin since every other plugin needs it

The algorithm should be solid. I didn't came up with a test case (only for the topological sort) but I tested it live on a server and seems to work.

thomasleveil commented:
nice, not a lot of code change in the end. I would have thought it would be more complicated to implement.

I have a question though: What happens if a plugin is disabled at run time (!disable command) and other plugins depend on it?
Should the dependent plugins also be disabled?

danielepantaleone commented:
Yes, also that part needs to be adapted but I wanted your opinions before changing more code

Inviato da iPhone

> Il giorno 09/mar/2015, alle ore 21:39, Thomas LÉVEIL <notifications@github.com> ha scritto:
> 
> nice, not a lot of code change in the end. I would have thought it would be more complicated to implement.
> 
> I have a question though: What happens if a plugin is disabled at run time (!disable command) and other plugins depend on it?
> Should the dependent plugins also be disabled?
> 
> —
> Reply to this email directly or view it on GitHub.
>

thomasleveil commented:
what about adding new events such as `EVT_PLUGIN_DISABLED` and `EVT_PLUGIN_ENABLED` so that dependent plugin could choose to react or not to a dependee plugin being disabled?

danielepantaleone commented:
we already have 2 hooks: onDisable and onEnable

danielepantaleone commented:
All should be done. I added also `EVT_PLUGIN_LOADED` and `EVT_PLUGIN_UNLOADED` and modified the plugin manager plugin to work the same way the parser does.

thomasleveil commented:
LGTM :+1:
